src/bin/mir-json-translate-libs.rs

```diff
@@ -0,0 +1,692 @@
+                    unit.target.src_path.strip_prefix(orig_pkg_dir).expect(
+                        "source file should be inside directory of Cargo.toml",
+                    );
+                let src_path = [
+                    custom_sources_dir.as_path(),
+                    Utf8Path::new(&unit.target.name),
+                    rel_src_path,
+                ]
+                .iter()
+                .collect();
+                let (linked_libs, linked_paths, cfgs, env) =
+                    match build_script_outputs.remove(&unit.pkg_id) {
+                        Some(bs) => {
+                            (bs.linked_libs, bs.linked_paths, bs.cfgs, bs.env)
+                        }
+                        None => (vec![], vec![], vec![], vec![]),
+                    };
+                CustomTarget::TargetLib(CustomTargetLib {
+                    crate_name: unit.target.name.as_str().into(),
+                    src_path,
+                    edition: unit.target.edition,
+                    linked_libs,
+                    linked_paths,
+                    // The build script's outputted cfgs does not include
+                    // features automatically computed by cargo, so we add them
+                    // in here. Technically we can get them either from the unit
+                    // graph or the artifact message, but we use the artifact
+                    // message version because it is closer to an actual
+                    // invocation of rustc. Not sure if there is actually a
+                    // difference.
+                    cfgs: artifact
+                        .features
+                        .iter()
+                        .map(|f| format!("feature=\"{f}\""))
+                        .chain(cfgs)
+                        .collect(),
+                    env,
+                })
+            }
+        },
+        dependencies: unit.dependencies,
+    };
+
+    let mut custom_graph = CustomUnitGraph {
+        units: unit_graph.units.into_iter().map(convert_unit).collect(),
+        roots: unit_graph.roots,
+    };
+
+    // Add our extra libraries to the unit graph.
+
+    let dep_core = custom_graph.get_unit_as_dep("core".into());
+    let dep_compiler_builtins =
+        custom_graph.get_unit_as_dep("compiler_builtins".into());
+    let dep_std = custom_graph.get_unit_as_dep("std".into());
+
+    // Add crucible
+    let dep_crucible = custom_graph.push_unit_as_dep(
+        CustomTargetLib {
+            crate_name: "crucible".into(),
+            src_path: [
+                custom_sources_dir.as_path(),
+                "crucible".into(),
+                "lib.rs".into(),
+            ]
+            .iter()
+            .collect(),
+            edition: Edition::E2021,
+            linked_libs: vec![],
+            linked_paths: vec![],
+            cfgs: vec![],
+            env: vec![],
+        },
+        vec![dep_compiler_builtins.clone(), dep_core.clone()],
+    );
```

Comment:
Adding `crucible` as a dependency is a surprisingly manual process. Is there any way to infer this information from `crucible`'s `Cargo.toml` file? (And similarly for `int512`, `byteorder`, etc.)

Reply:
Unfortunately only `byteorder` has a `Cargo.toml` out of those 4 libraries. Even if we added it for the rest, I think the only information we can easily get from `Cargo.toml` is `crate_name`, `src_path`, and `edition`. I think everything else would require Cargo to do a build, and since these aren't in the existing unit graph, we would have to generate separate unit graphs for them and merge them. Plus we still need to manually add `crucible` as a dependency of `alloc` anyways. So I think it's more trouble than it's worth.

Reply:
OK, fair enough. In that case, could you leave a comment somewhere describing how you figured out what metadata for use for these libraries? My hope is that if we have to do this again in the future (e.g., if we add another extra dependency for some reason), then we can rely on this comment to help guide us rather than figuring it out from scratch.

Reply:
How I figured it out was just by transcribing the `mir-json` commands in the old `translate_libs.sh` :smile:

I'm afraid I don't know how whoever wrote those figured it out in the first place...